### DISCUSSION TITLE
Trophies

### DISCUSSION THREAD

**Discussion by zwrawr** [ludumdare/ludumdare#2021] [Ideas]
>  (@jupiter-hadley)
    What a nice website. I do miss trophies though. </3 I included it in my Ludum Dare 38 compilation video series, if you’d like to take a look :)

https://youtu.be/5I8aoDLW98E?t=462

**Comment by mikekasprzak** [comment]
One of these days, I'll tell you all about the Trophy Shelf. :wink:

**Comment by Aurel300** [comment]
Perhaps could be a part of #807?

**Comment by mikekasprzak** [comment]
Okay, since I'm making reference to it (I forgot I didn't actually document it), here's some details on the trophy shelf.

On the user page, I'd like to see a few tabs. One tab is a chart view, where you can see how well your games have done across events. Another tab is the **Trophy Shelf**.

The trophy shelf is a bit of a fun feature. There are 2 classes of trophies:

* Event Trophies (given to you by placing in Ludum Dare events)
* User Trophies (given to you by users that want to highlight something special about your game)

As the name "Trophy Shelf" suggests, it's an actual digital shelf you can place and arrange your trophies on. Your personal trophy shelf can be arranged as you see fit. If you want to set them up neatly you can. If you want a chaotic mess, knocking them over, you can do that too. When I call this a fun feature, I'm referring to the simple 2D physics simulation that lets you pick-up and move them around (FYI 2D physics was my forte back in the day).

Generally speaking Event trophies will be large, and User trophies small. What's needed is some way to let other users like/agree with the trophy, and the more users that do, the larger the trophy becomes.

**NOTE**: In the interim, a **Trophy List** page on a user profile may suffice. The shelf is very much a fun "for later" concept.

Trophies in the past used to be a free-for-all, anyone can give anyone a trophy. Especially for people that didn't participate, this was the only way they could say a game was special. This concept isn't a bad one, like being able to say "game X is my favourite platformer". Coming from a YouTuber or website, this can still sound impressive.

But what might be nice is a process to, among the larger Ludum Dare group, create some Trophies that multiple users get behind. The problem is again, how that's achieved. There almost needs to be an extra phase to the voting. Say in the first 2 weeks of Ludum Dare, trophies and candidates are proposed, and during the last week actual votes are tallied for them. The problem though is the same problem with voting: how do we get enough people to cast votes towards the trophies?

Anyway, at the very least I wanted to blueprint-out the idea of multiple class of trohpies: winning events, individual trophies, and group trophies. Exactly how these all function is TBD.

**Comment by zwrawr** [comment]
In reference to the chart view, @pjnovas made ldstats. Which seems to be what mike wants on the site. It got updated to work for LD38 ( pjnovas/ldstats#11 ). 

We could see if @pjnovas would like to work on adding a similar feature to this site?

* http://ldstats.info/
* https://github.com/pjnovas/ldstats

**Comment by local-minimum** [comment]
When #1279 has been merged, I could work on this. My thinking is to do it something like the included image. First implementation would not allow you to organize your shelf, but show the trophies.
I'd like help with the backend for creating and getting trophies:

* Make new trophy (image, title, caption (?))
* Get all trophies created by a user
* Get all trophies awarded by user (most recent first).

![ld_trophies](https://user-images.githubusercontent.com/8041100/31818748-bd9461c8-b599-11e7-852a-1f87e8911c4e.png)

**Comment by sgstair** [comment]
Some ideas on implementation of this:
* Trophies will be nodes, of type (say) trophy/user (user defined) and trophy/event (for any autogenerated trophies in an event). As such, creating a Trophy doesn't require any new API (Just some minor changes to allow creation of trophy/user nodes)
  * Trophy image will be a meta on the trophy node. Title will be the name of the trophy. Body will be maybe a longer description
  * Parent will be the user's game node (for trophy/user) or the event for trophy/event - users with game nodes have implicit permission to create nodes of trophy/user
* Awarding / Accepting a trophy will be facilitated with new API. API ownership of user's trophy list is necessary to prevent people from gaming the system.
* Awarded trophies will raise notifications (probably type "trophy", with node id = trophy node)
* Awarded trophies may be private user meta, and accepted trophies may be public user meta
  * I'm thinking of a scheme like "trophy-[node id]" = "1" for each present trophy.
  * My concern with using meta is with a lot of trophies this clutters the user object. Using an alternate query path for trophies might be ideal. Visible badges (which might be implemented through the same path as trophies) would probably be meta as we expect them to be visible in comments, Trophies less so.

Let me know if I'm missing cases here, or if you see better solutions.

**Comment by local-minimum** [comment]
To be clear, 'trophy/user' parent will be trophy creating/giving, not trophy recieving user.

As for cluttering, I think this is true, but also 'trophy/event' should be expected to superseed how category winner trophies are currently handled. So we need to think about that too. A game/item might be the reciever rather than user but user should also show the same trophies.

**Comment by mikekasprzak** [comment]
I'm going to have to come back to this (I'm not fully up-to-date on what's all here), but just for context I don't want the old trophy system, I want a better new system.

One of the concerns with the old system is the freeform nature to use any bit of artwork for a trophy. It is a problem when folks can decide to use a picture of Super Mario for a trophy, because we legally can't use that. In the past we let stuff like this pass, but in the future we need to be a bit more responsible.
